embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpansRepository.kt

```diff
@@ -0,0 +1,62 @@
+package io.embrace.android.embracesdk.internal.spans
+
+import io.embrace.android.embracesdk.spans.EmbraceSpan
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Allows the tracking of [EmbraceSpan] instances so that their references can be retrieved with its associated spanId
+ */
+internal class SpansRepository {
+    private val activeSpans: MutableMap<String, EmbraceSpan> = ConcurrentHashMap()
+    private val completedSpans: MutableMap<String, EmbraceSpan> = ConcurrentHashMap()
+
+    /**
+     * Track the [EmbraceSpan] if it has been started and it's not already tracked.
+     */
+    fun started(embraceSpan: EmbraceSpan) {
+        embraceSpan.spanId?.let { spanId ->
+            if (activeSpans[spanId] == null && completedSpans[spanId] == null) {
+                synchronized(activeSpans) {
+                    if (activeSpans[spanId] == null && completedSpans[spanId] == null) {
+                        if (embraceSpan.isRecording) {
+                            activeSpans[spanId] = embraceSpan
+                        } else {
+                            completedSpans[spanId] = embraceSpan
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Transition active span to completed span if the span is tracked and the span is actually stopped.
+     */
+    fun stopped(spanId: String) {
+        synchronized(activeSpans) {
+            activeSpans[spanId]?.takeIf { !it.isRecording }?.let { activeSpans.remove(spanId) }?.let { span ->
```

Comment:
Is the span guaranteed to always be `false` for `isRecording` at this point?

Reply:
It's not if someone make a mistake and calls this on a span that isn't actually stopped. This provides another level of validation.